File: go/plugins/boru/boru.go
// The boru plugin: a boru vault through its CLI, or over `boru vault
// serve`. Needs a child process, or HTTPS in wire mode. A port of
// typescript/plugins/boru.ts.
package boru

import (
	"bytes"
	"errors"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"strings"

	"github.com/voxgig/sekreto/go/plugins/httpjson"
	"github.com/voxgig/sekreto/go/sekreto"
)

// Provider reads a boru vault (https://github.com/boru-lang/boru).
//
// Two ways in, both boru's own.
//
// With no Addr, the CLI: `boru vault get --reveal <alias>` prints the
// secret on stdout, and nothing else. The passphrase is read by boru itself
// from BORU_VAULT_PASSPHRASE; sekreto never accepts it as config and never
// puts it on a command line, where it would show up in the process table.
//
// With an Addr, boru's wire protocol: `boru vault serve` publishes a
// read-only, HashiCorp-shaped provision API (boru's
// design/VAULT-WIRE-PROTOCOL.0.md), authenticated by a capability token
// from `boru vault grant`. A sekreto name is already a valid boru alias,
// and boru aliases keep their dots, so api.token is the single path segment
// api.token - not the api/token split a HashiCorp KV gets. The value is the
// `value` field. A 404 is a miss; anything else the server refuses (a
// revoked capability, a sealed vault) is an error.
//
// boru's `vault proxy` and `vault mcp` remain out of bounds: they are a
// credential *broker*, built precisely so the caller never receives the
// credential. `vault serve` is the provision endpoint, built to hand the
// value back - that is the one sekreto uses.
type Provider struct {
	Command   string
	Namespace string
	Home      string
	Addr      string
	Token     string
	Mount     string
}

func (provider *Provider) command() string {
	if "" == provider.Command {
		return "boru"
	}
	return provider.Command
}

func (provider *Provider) wireaddr() string {
	return strings.TrimSuffix(provider.Addr, "/")
}

func (provider *Provider) wiremount() string {
	if "" == provider.Mount {
		return "secret"
	}
	return provider.Mount
}

func (provider *Provider) wirelookup(name string) (string, bool, error) {
	addr := provider.wireaddr()
	if err := sekreto.CheckAddr(addr); nil != err {
		return "", false, err
	}

	// The dotted name stays one segment: boru aliases keep their dots.
	alias := name
	if "" != provider.Namespace {
		alias = provider.Namespace + "/" + name
	}
	target := addr + "/v1/" + provider.wiremount() + "/data/" + alias

	status, body, err := httpjson.Get(target, map[string]string{"X-Vault-Token": provider.Token})
	if nil != err {
		return "", false, err
	}

	if http.StatusNotFound == status {
		return "", false, nil
	}

	if http.StatusOK != status {
		return "", false, sekreto.Fail("sekreto: boru serve error: " + strconv.Itoa(status) + ": " + target)
	}

	value := httpjson.Dig(body, "data", "data", "value")
	if nil == value {
		return "", false, nil
	}

	return httpjson.ToString(value), true, nil
}

func (provider *Provider) Lookup(name string) (string, bool, error) {
	if err := sekreto.CheckName(name); nil != err {
		return "", false, err
	}

	if "" != provider.Addr {
		return provider.wirelookup(name)
	}

	alias := name
	if "" != provider.Namespace {
		alias = provider.Namespace + ":" + name
	}

	run := exec.Command(provider.command(), "vault", "get", "--reveal", alias)

	if "" != provider.Home {
		run.Env = append(os.Environ(), "BORU_HOME="+provider.Home)
	}

	var out, errout bytes.Buffer
	run.Stdout = &out
	run.Stderr = &errout

	err := run.Run()

	if nil == err {
		// boru prints the value and one newline, and nothing else.
		return strings.TrimSuffix(out.String(), "\n"), true, nil
	}

	var exiterr *exec.ExitError
	if !errors.As(err, &exiterr) {
		return "", false, sekreto.Fail("sekreto: cannot run " + provider.command() + ": " + err.Error())
	}

	why := strings.TrimSpace(errout.String())

	// "no alias named" is boru saying it does not hold this secret, which is a
	// miss: the chain carries on to the next provider. A locked vault or a
	// wrong passphrase is not a miss - treating it as one would fall through
	// to a weaker store without saying so.
	if borumiss(why) {
		return "", false, nil
	}

	if "" == why {
		why = "exit " + strconv.Itoa(exiterr.ExitCode())
	}

	return "", false, sekreto.Fail("sekreto: boru vault error: " + why)
}

func (provider *Provider) Describe() string {
	if "" != provider.Addr {
		return "boru:" + provider.wireaddr()
	}
	if "" != provider.Namespace {
		return "boru:" + provider.Namespace
	}
	return "boru"
}

// borumiss reports whether a boru failure means "no such secret" rather than
// "I could not answer". Matched on boru's own wording for a missing alias.
func borumiss(why string) bool {
	return strings.Contains(why, "no alias named")
}

// Plugin is the `boru` provider kind, as a voxgig/plugin definition.
var Plugin = sekreto.ProviderPlugin("boru", func(spec *sekreto.ProviderSpec) (sekreto.Provider, error) {
	return &Provider{
		Command:   spec.Command,
		Namespace: spec.Namespace,
		Home:      spec.Home,
		Addr:      spec.Addr,
		Token:     spec.Token,
		Mount:     spec.Mount,
	}, nil
})

File: lua/src/sekreto/plugins/boru.lua
-- The boru plugin: a boru vault, through its own CLI or its own wire
-- protocol. Needs a child process, and HTTPS for the wire path.
--
-- A port of typescript/plugins/boru.ts, which is canonical.

local err = require('sekreto.err')
local name = require('sekreto.name')
local addr = require('sekreto.addr')
local providers = require('sekreto.providers')
local json = require('sekreto.plugins.json')
local httpjson = require('sekreto.plugins.httpjson')
local support = require('sekreto.plugins.support')
local net = require('sekreto.plugins.net')

local fail = err.fail
local nonempty = providers.nonempty
local trimslash = providers.trimslash
local providerplugin = providers.providerplugin
local stripnewline = support.stripnewline
local fetchjson = httpjson.fetchjson
local runcmd = net.runcmd

local M = {}

--- Does this boru failure mean "no such secret" rather than "I could not
--- answer"? Matched on boru's own wording for a missing alias.
function M.borumiss(why)
  return nil ~= why:find('no alias named', 1, true)
end

--- A boru vault (https://github.com/boru-lang/boru).
---
--- Two ways in, both boru's own. With no `addr`, the CLI:
--- `boru vault get --reveal <alias>` prints the secret on stdout and
--- nothing else. The passphrase is read by boru itself from
--- BORU_VAULT_PASSPHRASE; sekreto never accepts it as config and never
--- puts it on a command line, where the process table would publish it.
---
--- With an `addr`, boru's wire protocol: a read-only, HashiCorp-shaped
--- provision API. A sekreto name is already a valid boru alias, and boru
--- aliases keep their dots, so `api.token` is the single path segment
--- `api.token` - not the `api`/`token` split a HashiCorp KV gets.
local function boru(command, namespace, home, useaddr, token, mount)
  local usecommand = nonempty(command) and command or 'boru'
  local address = (nil == useaddr) and '' or trimslash(useaddr)
  local usetoken = token or ''
  local usemount = nonempty(mount) and mount or 'secret'

  local function wirelookup(secret)
    addr.checkaddr(address)

    local alias = nonempty(namespace) and (namespace .. '/' .. secret) or secret
    local url = address .. '/v1/' .. usemount .. '/data/' .. alias

    local res = fetchjson('GET', url, { { 'X-Vault-Token', usetoken } })

    if 404 == res.status then
      return nil
    end

    if 200 ~= res.status then
      fail('sekreto: boru serve error: ' .. res.status .. ': ' .. url)
    end

    return json.text(json.dig(res.body, 'data', 'data', 'value'))
  end

  return {
    lookup = function(secret)
      name.checkname(secret)

      if '' ~= address then
        return wirelookup(secret)
      end

      local alias = nonempty(namespace) and (namespace .. ':' .. secret) or secret

      -- BORU_HOME is the only variable sekreto sets for the child; the
      -- passphrase is boru's own to read from BORU_VAULT_PASSPHRASE, and
      -- is never config here and never on a command line.
      local overrides = nil
      if nonempty(home) then
        overrides = { 'BORU_HOME=' .. home }
      end

      local out, why, status = runcmd(
        { usecommand, 'vault', 'get', '--reveal', alias }, overrides, usecommand
      )

      if 0 == status then
        -- boru prints the value and one newline, and nothing else.
        return stripnewline(out)
      end

      -- "no alias named" is boru saying it does not hold this secret,
      -- which is a miss. A locked vault or a wrong passphrase is not -
      -- treating it as one would fall through to a weaker store without
      -- saying so.
      if M.borumiss(why) then
        return nil
      end

      fail('sekreto: boru vault error: ' ..
        (('' == why) and ('exit ' .. status) or why))
    end,
    describe = function()
      if '' ~= address then
        return 'boru:' .. address
      end
      return 'boru' .. (nonempty(namespace) and (':' .. namespace) or '')
    end,
  }
end

--- The kind, as a voxgig/plugin definition.
M.boru = providerplugin('boru', function(spec)
  return boru(spec.command, spec.namespace, spec.home, spec.addr,
    spec.token, spec.mount)
end)

return M
